vhdx/src/vhdx_io.rs

```diff
@@ -0,0 +1,235 @@
+// Copyright © 2021 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::vhdx_bat::{self, BatEntry};
+use crate::vhdx_metadata::{self, DiskSpec};
+use remain::sorted;
+use std::fmt::{self, Display};
+use std::fs::File;
+use std::io::{self, Read, Seek, SeekFrom, Write};
+
+const SECTOR_SIZE: u64 = 512;
+
+#[sorted]
+#[derive(Debug)]
+pub enum Error {
+    // Invalid BAT entry state
+    InvalidBatEntryState,
+    // Invalid number of BAT entries
+    InvalidBatIndex,
+    // Resulting too big disk size
+    InvalidDiskSize,
+    // Failed reading a sector block
+    ReadSectorBlock(io::Error),
+    // Failed resizing the VHDx file
+    ResizeFile(io::Error),
+    // Differencing mode is not supported
+    UnsupportedMode,
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+impl Display for Error {
+    #[remain::check]
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use self::Error::*;
+
+        #[sorted]
+        match self {
+            InvalidBatEntryState => write!(f, "Invalid BAT entry state"),
+            InvalidBatIndex => write!(f, "Invalid BAT entry count"),
+            InvalidDiskSize => write!(f, "Invalid disk size"),
+            ReadSectorBlock(e) => write!(f, "Failed reading sector blocks form file {}", e),
+            ResizeFile(e) => write!(f, "Failed changing file length {}", e),
+            UnsupportedMode => write!(f, "Differencing mode is not supported yet"),
+        }
+    }
+}
+
+macro_rules! round_up {
+    ($n:expr, $align:expr) => {{
+        if $align > $n {
+            $align
+        } else {
+            let rem = $n % $align;
+            (($n / $align) + rem) * $align
+        }
+    }};
+}
+
+#[derive(Default)]
+struct Sector {
+    bat_index: u64,
+    free_sectors: u64,
+    free_bytes: u64,
+    file_offset: u64,
+    block_offset: u64,
+}
+
+impl Sector {
+    /// Translate secotor index and count of data in file to actual offsets and BAT index
+    pub fn new(
+        disk_spec: &DiskSpec,
+        bat: &[BatEntry],
+        sector_index: &u64,
+        sector_count: &u64,
+    ) -> Result<Sector> {
+        let mut sector = Sector::default();
+
+        sector.bat_index = *sector_index / disk_spec.sectors_per_block as u64;
+        sector.block_offset = *sector_index % disk_spec.sectors_per_block as u64;
+        sector.free_sectors = disk_spec.sectors_per_block as u64 - sector.block_offset;
+        if sector.free_sectors > *sector_count {
```

Comment:
Shouldn't this be an error? I mean in case the calculated `free_sectors` is bigger than what is expected, that sounds like we should throw an error.